designs/osc_containerization/OSC-containerization.md

```diff
@@ -0,0 +1,90 @@
+
+## Assignees
+Arvind Nadendla https://github.com/arvindn05
+
+## Background
+
+OSC containerization for deployment allows greater flexibility on mixing services on same host.
+
+For example, in case of openstack we can co-locate osc on the controller node with the other controller services like keystone, nova etc. We can follow the same lifecycle as those services and utilize similar mechanisms for upgrades, data persistence, HA etc
+
+It also simplifies packaging and distribution.
+
+## Constraints and Assumptions
+
+##### Constraint
+- Support migrating current OSC installation to container based installation.
+
+##### Assumption
+- OSC will be deployed as a container only. The Virtual Machine appliance model will not be supported going forward.
+- For migration of VM-based OSC to container based OSC, some manual steps maybe required. Scripts should be provided in this case.
+
+## Design Changes
+
+##### Persistent data
+Persistent data stored by OSC will need to be decoupled and managed outside the lifecycle of the container.
+
+The following data will need to be decoupled from the lifecycle of the OSC container:
+- H2 database
+- Keystore
+- Truststore
+- Plugins
+- Service Function catalog image files
+- Log file
+- vmidcServer.conf
+
+This data would need to be exposed in docker volumes to the OSC container.
+
+##### Life cycle management
+Remove restart and shutdown operations possbile via OSC UI
+Remove inplace upgrade of OSC via server upgrade bundles
+
+##### CLI/Restricted Shell
+Remove restricted shell. Restricted shell is no longer applicable as we will be using standard base containers.
+Note: It is technically feasible to have a restricted shell if we build up the OSC container image from scratch. For now, this is out of scope.
+
+##### Network/Ip address management
+Remove network management functionality. Setting custom IP address within a docker container does not make sense as the IP is not exposed outside of the container.
+
+We can expose OSC endpoints as a virtual ip from the host or use different ports to expose OSC
+
+### POC Implementation
+
+###### Scope
+- Create docker files for OSC
+- Use Docker volume to store persistent data and decouple from OSC code
+	- Moved log files to persistent volume
+	- Moved Database to persisent volume
+	- Moved Plugins to persistent volume
+- Launch different OSC containers and make sure they are using the same persistent data
+
+Code is available [here](https://github.com/arvindn05/osc-core/tree/docker)
+
+### REST API
+N/A
+
+### OSC SDKs
+
+#### VNF Security Manager SDK
+N/A
+
+#### SDN Controller SDK
+N/A
+
+### OSC Entities
+N/A
+
+### OSC UI
+N/A
+
+### OSC Synchronization Tasks
+N/A
+
+## Tests
+N/A
+
+## References
+### [Dockerfile reference](https://docs.docker.com/engine/reference/builder/)
```

Comment:
Can you add more references and we don't have to search for additional information elsewhere?

Reply:
Any examples of what other information you want here? A lot of this functionality is documented as part of docker or other tutorials online.